pkg/apis/core/v1alpha1/kubefedconfig_types.go

```diff
@@ -25,59 +25,98 @@ import (
-	ClusterHealthCheck ClusterHealthCheckConfig `json:"cluster-health-check,omitempty"`
-	SyncController     SyncControllerConfig     `json:"sync-controller,omitempty"`
+	Scope apiextv1b1.ResourceScope `json:"scope"`
+	// +optional
+	ControllerDuration *DurationConfig `json:"controllerDuration,omitempty"`
+	// +optional
+	LeaderElect *LeaderElectConfig `json:"leaderElect,omitempty"`
+	// +optional
+	FeatureGates []FeatureGatesConfig `json:"featureGates,omitempty"`
+	// +optional
+	ClusterHealthCheck *ClusterHealthCheckConfig `json:"clusterHealthCheck,omitempty"`
+	// +optional
+	SyncController *SyncControllerConfig `json:"syncController,omitempty"`
 }
 
 type DurationConfig struct {
 	// Time to wait before reconciling on a healthy cluster.
-	AvailableDelay metav1.Duration `json:"available-delay,omitempty"`
+	// +optional
+	AvailableDelay metav1.Duration `json:"availableDelay,omitempty"`
 	// Time to wait before giving up on an unhealthy cluster.
-	UnavailableDelay metav1.Duration `json:"unavailable-delay,omitempty"`
+	// +optional
+	UnavailableDelay metav1.Duration `json:"unavailableDelay,omitempty"`
 }
 type LeaderElectConfig struct {
 	// The duration that non-leader candidates will wait after observing a leadership
 	// renewal until attempting to acquire leadership of a led but unrenewed leader
 	// slot. This is effectively the maximum duration that a leader can be stopped
 	// before it is replaced by another candidate. This is only applicable if leader
 	// election is enabled.
-	LeaseDuration metav1.Duration `json:"lease-duration,omitempty"`
+	// +optional
+	LeaseDuration metav1.Duration `json:"leaseDuration,omitempty"`
 	// The interval between attempts by the acting master to renew a leadership slot
 	// before it stops leading. This must be less than or equal to the lease duration.
 	// This is only applicable if leader election is enabled.
-	RenewDeadline metav1.Duration `json:"renew-deadline,omitempty"`
+	// +optional
+	RenewDeadline metav1.Duration `json:"renewDeadline,omitempty"`
 	// The duration the clients should wait between attempting acquisition and renewal
 	// of a leadership. This is only applicable if leader election is enabled.
-	RetryPeriod metav1.Duration `json:"retry-period,omitempty"`
+	// +optional
+	RetryPeriod metav1.Duration `json:"retryPeriod,omitempty"`
 	// The type of resource object that is used for locking during
 	// leader election. Supported options are `configmaps` (default) and `endpoints`.
-	ResourceLock string `json:"resource-lock,omitempty"`
+	// +optional
+	ResourceLock ResourceLockType `json:"resourceLock,omitempty"`
 }
+
+type ResourceLockType string
+
+const (
+	ConfigMapResourceLock ResourceLockType = "configmaps"
+	EndpointResourceLock  ResourceLockType = "endpoints"
+)
+
 type FeatureGatesConfig struct {
-	Name    string `json:"name,omitempty"`
-	Enabled bool   `json:"enabled,omitempty"`
+	Name          string            `json:"name"`
+	Configuration ConfigurationMode `json:"configuration"`
```

Comment:
(No action required) I get that you're trying to be consistent, but this seems odiously excessive to me.  The chances of us having to extend the way we enable a feature gate or not is nil.  We're more likely to bin the very idea of feature gates - it's a terrible way to manage GA features.